Add support to create a DriveItem with many attributes in a single request

lucasjrt opened the pull request:
## Introduction
This lib does not have the ability to create a `DriveItem` with a description or any other fields but name, at least I couldn't find a way to do this in a single request.

Probably it is possible to add a description to a `DriveItem` with `update_item`/`update_item_with_option`, but it is inefficient to perform two requests when a single one can be done.

## Changes
- Introduce a few good-to-have implementations for `ConflictBehavior` enum
- Update the DriveItem struct macro to allow the conflict behavior to be set
- Refactor the original `create_folder_with_options` to use the new implementation

## Impact and benefits
- Users can now create DriveItems with additional fields, such as description, in a single request, improving usability and efficiency.
- The changes do not introduce breaking changes and are backward-compatible with existing usage of the library.
- Any supported field can be added to the request, as long as it is compatible with the [Microsoft documentation](https://learn.microsoft.com/en-us/graph/api/resources/driveitem?view=graph-rest-1.0#properties).

## Tests
Tests are already covered  as `create_document_with_options` is using the new implementation.

## Requests for review
Feedback and review from other contributors and maintainers are highly appreciated, especially regarding the implementation details, and documentation clarity.

lucasjrt committed:
Add `create_drive_item` with dynamic parameters

oxalica commented on src/lib.rs:
```diff
@@ -98,6 +98,32 @@ pub enum ConflictBehavior {
+        }
+    }
+}
+
```
Why do we need these? Doesn't deriving `Serialize` and `Deserialize` enough?

lucasjrt replied:
My idea was that
```
conflict_behavior.to_string()
```

would be more readable than
```
json!(conflict_behavior)
    .as_str()
    .expect("serialization not to fail")
    .to_string()
```

But I agree it's better to work with serializing, it's easier to maintain. I'll make a new commit with the changes.

oxalica commented on src/onedrive.rs:
```diff
@@ -269,6 +270,39 @@ impl OneDrive {
+    pub async fn create_drive_item<'a, P, D>(
+        &self,
+        parent_item: P,
+        drive_item: D,
+    ) -> Result<DriveItem>
```
The conflict behavior is a control option of the method instead of a property of a `DriveItem`, thus it makes more sense to add it here as a parameter rather than inside `DriveItem` struct. 
`#[serde(flatten)]` can be used to merge two structs during serialization. You can follow and/or reuse the impl of [`new_upload_session_with_initial_option`](https://github.com/oxalica/onedrive-api/blob/9c9516818934f4ed79394ee46c0776d5725943f3/src/onedrive.rs#L452).

```suggestion
        drive_item: D,
        conflict_behavior: ConflictBehavior,
    ) -> Result<DriveItem>
```

Also I think we should assert that `drive_item` has `folder` set, or it's a misuse and it will fail anyway.

lucasjrt replied:
I agree, it makes sense to extract it from the struct.

Regarding the assertion of having `folder` set, I disagree as the `DriveItem` may have another facet. See the comment I've added talking about the facets [here](https://github.com/oxalica/onedrive-api/pull/6#issuecomment-1935800548). We can check that at least one of those four facets is set though.

oxalica commented on src/resource.rs:
```diff
@@ -255,7 +255,19 @@ define_resource_object! {
+        /// [get_item]: ../struct.OneDrive.html#method.get_item
+        /// [get_item_with_opt]: ../struct.OneDrive.html#method.get_item_with_option
+        [unselectable]
+        pub conflict_behavior @"@microsoft.graph.conflictBehavior": Option<String>,
```
I don't think we need this. See the explanation above.

oxalica reviewed (changes requested):
Thanks for the PR! It's good to have this API for folders, and we already had a similar `new_upload_session_with_initial_option` for creating files with given `DriveItem` properties.

I'm not really sure about the naming. I was thinking about `create_folder_with_initial_option`, but I have to admit this name is kinda awkward and long. `create_drive_item` is simpler and follows the upstream API naming, but it doesn't reflect that this API can only be used to create folders, not files.

lucasjrt commented:
@oxalica, it's my pleasure to contribute!

The idea of `create_drive_item` is to allow the user to create any generic `DriveItem` it may desire, not only folders. Basically a `DriveItem` itself is not a valid structure if it doesn't contain at least one _facet_, and these are not only limited to folder or file. Here's an example of an attempt to create a `DriveItem` without any facet:
```
Error creating file: Api error with 400 Bad Request: (invalidRequest) Must provide one of the following facets to create an item: Bundle, File, Folder, RemoteItem
```


You can see the available facets [here](https://learn.microsoft.com/en-us/onedrive/developer/rest-api/resources/?view=odsp-graph-online#facets).

Let's say someone has a use case that they need to create a Bundle or a RemoteItem instead of a folder with description (my current use case). It's convenient to have the flexibility of `create_drive_item` and allow the definition of any facet than limiting to pre-defined implementations, which are also good to have, but not as the only option.

I just looked at `new_upload_session_with_initial_option`, and it's quite similar, but it returns an `UploadSession` and `UploadSessionMeta` instead of the created `DriveItem`, so let's say if one is interested int the ID of the new item, it will be required to make a new request to get it.

oxalica commented on src/onedrive.rs:
```diff
@@ -269,56 +270,83 @@ impl OneDrive {
-        name: &FileName,
-        option: DriveItemPutOption,
+        drive_item: DriveItem,
+        conflict_behavior: Option<ConflictBehavior>,
```
I  noticed that currently `create_folders*` uses `DriveItemPutOption` which includes additional `If-{Non-,}Match` support. If that's valid, we should go with:

```suggestion
        option: DriveItemPutOption,
```
and `.apply(option)` on the request builder.

lucasjrt replied:
Done

oxalica commented on src/onedrive.rs:
```diff
@@ -269,56 +270,83 @@ impl OneDrive {
     /// # See also
-    /// [Microsoft Docs](https://docs.microsoft.com/en-us/graph/api/driveitem-post-children?view=graph-rest-1.0)
     ///
+    /// [Microsoft Docs](https://learn.microsoft.com/en-us/graph/api/resources/driveitem?view=graph-rest-1.0)
```
Please use this link mentioning `POST` method for item creation instead of the docs for `DriveItem` itself:
<https://learn.microsoft.com/en-us/onedrive/developer/rest-api/api/driveitem_post_children>

 It's also better to mention the difference between `create_folders*`: additional item properties can be specified, and facets like `folders` must be provided manually or it will fail.

lucasjrt replied:
Done

oxalica reviewed (changes requested):
> The idea of create_drive_item is to allow the user to create any generic DriveItem it may desire, not only folders. Basically a DriveItem itself is not a valid structure if it doesn't contain at least one facet, and these are not only limited to folder or file. 

Thanks for the explanation. It's good to mention that in the docs of the new method.

It seems GitHub has trouble rendering the diff. Could you squash all commits?

lucasjrt commented:
I've squashed the commits, updated some docs, added options like the other methods and added a validation: if none of the required facets are set, the code will fail instead of sending a request to the server. Not sure if this is a good thing, but I've added just in case. Let me know what you think.

oxalica commented on src/onedrive.rs:
```diff
@@ -269,56 +270,101 @@ impl OneDrive {
+        drive_item: DriveItem,
         option: DriveItemPutOption,
     ) -> Result<DriveItem> {
+        if drive_item.bundle.is_none()
+            && drive_item.file.is_none()
+            && drive_item.folder.is_none()
+            && drive_item.remote_item.is_none()
+        {
+            // Locally preventing a Bad Request. Other errors are still possible
+            return Err(Error::from_error_response(
+                reqwest::StatusCode::BAD_REQUEST,
+                ErrorResponse {
+                    code: "invalidRequest".to_string(),
+                    message: "Must provide one of the following facets to create an item: Bundle, File, Folder, RemoteItem".to_string(),
+                    inner_error: None,
+                },
+            ));
+        }
```
My intent was to add `assert!` to panic immediately because it's clearly a coding error, rather than something that *might* happen during a normal workflow as an `Err` return.

Well, in the other hand, hard-checking here assumes our check is the same as the upstream restrictions. But there may be more restrictions or new facets added in the future.

Thus I think it may be better to just remove these check and let upstream decide. This *should* be enough since the documentation already mentions it. Sorry for the inconvenience.

lucasjrt replied:
I agree, and don't worry. I've removed the code already, so please let me know of any other changes you might see to improve the code.

oxalica reviewed (changes requested):
Otherwise LGTM.

oxalica commented:
Thanks!
